// @formatjs/icu-messageformat-parser/manipulator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hoistSelectors = hoistSelectors;
exports.isStructurallySame = isStructurallySame;
var tslib_1 = require("tslib");
var types_1 = require("./types");
function cloneDeep(obj) {
    if (Array.isArray(obj)) {
        // @ts-expect-error meh
        return tslib_1.__spreadArray([], obj.map(cloneDeep), true);
    }
    if (obj !== null && typeof obj === 'object') {
        // @ts-expect-error meh
        return Object.keys(obj).reduce(function (cloned, k) {
            // @ts-expect-error meh
            cloned[k] = cloneDeep(obj[k]);
            return cloned;
        }, {});
    }
    return obj;
}
function hoistPluralOrSelectElement(ast, el, positionToInject) {
    // pull this out of the ast and move it to the top
    var cloned = cloneDeep(el);
    var options = cloned.options;
    cloned.options = Object.keys(options).reduce(function (all, k) {
        var newValue = hoistSelectors(tslib_1.__spreadArray(tslib_1.__spreadArray(tslib_1.__spreadArray([], ast.slice(0, positionToInject), true), options[k].value, true), ast.slice(positionToInject + 1), true));
        all[k] = {
            value: newValue,
        };
        return all;
    }, {});
    return cloned;
}
function isPluralOrSelectElement(el) {
    return (0, types_1.isPluralElement)(el) || (0, types_1.isSelectElement)(el);
}
function findPluralOrSelectElement(ast) {
    return !!ast.find(function (el) {
        if (isPluralOrSelectElement(el)) {
            return true;
        }
        if ((0, types_1.isTagElement)(el)) {
            return findPluralOrSelectElement(el.children);
        }
        return false;
    });
}
/**
 * Hoist all selectors to the beginning of the AST & flatten the
 * resulting options. E.g:
 * "I have {count, plural, one{a dog} other{many dogs}}"
 * becomes "{count, plural, one{I have a dog} other{I have many dogs}}".
 * If there are multiple selectors, the order of which one is hoisted 1st
 * is non-deterministic.
 * The goal is to provide as many full sentences as possible since fragmented
 * sentences are not translator-friendly
 * @param ast AST
 */
function hoistSelectors(ast) {
    for (var i = 0; i < ast.length; i++) {
        var el = ast[i];
        if (isPluralOrSelectElement(el)) {
            return [hoistPluralOrSelectElement(ast, el, i)];
        }
        if ((0, types_1.isTagElement)(el) && findPluralOrSelectElement([el])) {
            throw new Error('Cannot hoist plural/select within a tag element. Please put the tag element inside each plural/select option');
        }
    }
    return ast;
}
function isStructurallySamePluralOrSelect(el1, el2) {
    var options1 = el1.options;
    var options2 = el2.options;
    if (Object.keys(options1).length !== Object.keys(options2).length) {
        return false;
    }
    for (var key in options1) {
        if (!options2[key]) {
            return false;
        }
        if (!isStructurallySame(options1[key].value, options2[key].value)) {
            return false;
        }
    }
    return true;
}
function isStructurallySame(a, b) {
    var aWithoutLiteral = a.filter(function (el) { return !(0, types_1.isLiteralElement)(el); });
    var bWithoutLiteral = b.filter(function (el) { return !(0, types_1.isLiteralElement)(el); });
    if (aWithoutLiteral.length !== bWithoutLiteral.length) {
        return false;
    }
    var elementsMapInA = aWithoutLiteral.reduce(function (all, el) {
        if ((0, types_1.isPoundElement)(el)) {
            all['#'] = el;
            return all;
        }
        all[el.value] = el;
        return all;
    }, {});
    var elementsMapInB = bWithoutLiteral.reduce(function (all, el) {
        if ((0, types_1.isPoundElement)(el)) {
            all['#'] = el;
            return all;
        }
        all[el.value] = el;
        return all;
    }, {});
    for (var _i = 0, _a = Object.keys(elementsMapInA); _i < _a.length; _i++) {
        var varName = _a[_i];
        var elA = elementsMapInA[varName];
        var elB = elementsMapInB[varName];
        if (!elB) {
            return false;
        }
        if (elA.type !== elB.type) {
            return false;
        }
        if ((0, types_1.isLiteralElement)(elA) || (0, types_1.isLiteralElement)(elB)) {
            continue;
        }
        if ((0, types_1.isArgumentElement)(elA) &&
            (0, types_1.isArgumentElement)(elB) &&
            elA.value !== elB.value) {
            return false;
        }
        if ((0, types_1.isPoundElement)(elA) || (0, types_1.isPoundElement)(elB)) {
            continue;
        }
        if ((0, types_1.isDateElement)(elA) ||
            (0, types_1.isTimeElement)(elA) ||
            (0, types_1.isNumberElement)(elA) ||
            (0, types_1.isDateElement)(elB) ||
            (0, types_1.isTimeElement)(elB) ||
            (0, types_1.isNumberElement)(elB)) {
            if (elA.value !== elB.value) {
                return false;
            }
        }
        if ((0, types_1.isPluralElement)(elA) &&
            (0, types_1.isPluralElement)(elB) &&
            !isStructurallySamePluralOrSelect(elA, elB)) {
            return false;
        }
        if ((0, types_1.isSelectElement)(elA) &&
            (0, types_1.isSelectElement)(elB) &&
            isStructurallySamePluralOrSelect(elA, elB)) {
            return false;
        }
        if ((0, types_1.isTagElement)(elA) && (0, types_1.isTagElement)(elB)) {
            if (elA.value !== elB.value) {
                return false;
            }
            if (!isStructurallySame(elA.children, elB.children)) {
                return false;
            }
        }
    }
    return true;
}
