// @formatjs/icu-messageformat-parser/lib/manipulator.js
import { __spreadArray } from "tslib";
import { isArgumentElement, isDateElement, isLiteralElement, isNumberElement, isPluralElement, isPoundElement, isSelectElement, isTagElement, isTimeElement, } from './types';
function cloneDeep(obj) {
    if (Array.isArray(obj)) {
        // @ts-expect-error meh
        return __spreadArray([], obj.map(cloneDeep), true);
    }
    if (obj !== null && typeof obj === 'object') {
        // @ts-expect-error meh
        return Object.keys(obj).reduce(function (cloned, k) {
            // @ts-expect-error meh
            cloned[k] = cloneDeep(obj[k]);
            return cloned;
        }, {});
    }
    return obj;
}
function hoistPluralOrSelectElement(ast, el, positionToInject) {
    // pull this out of the ast and move it to the top
    var cloned = cloneDeep(el);
    var options = cloned.options;
    cloned.options = Object.keys(options).reduce(function (all, k) {
        var newValue = hoistSelectors(__spreadArray(__spreadArray(__spreadArray([], ast.slice(0, positionToInject), true), options[k].value, true), ast.slice(positionToInject + 1), true));
        all[k] = {
            value: newValue,
        };
        return all;
    }, {});
    return cloned;
}
function isPluralOrSelectElement(el) {
    return isPluralElement(el) || isSelectElement(el);
}
function findPluralOrSelectElement(ast) {
    return !!ast.find(function (el) {
        if (isPluralOrSelectElement(el)) {
            return true;
        }
        if (isTagElement(el)) {
            return findPluralOrSelectElement(el.children);
        }
        return false;
    });
}
/**
 * Hoist all selectors to the beginning of the AST & flatten the
 * resulting options. E.g:
 * "I have {count, plural, one{a dog} other{many dogs}}"
 * becomes "{count, plural, one{I have a dog} other{I have many dogs}}".
 * If there are multiple selectors, the order of which one is hoisted 1st
 * is non-deterministic.
 * The goal is to provide as many full sentences as possible since fragmented
 * sentences are not translator-friendly
 * @param ast AST
 */
export function hoistSelectors(ast) {
    for (var i = 0; i < ast.length; i++) {
        var el = ast[i];
        if (isPluralOrSelectElement(el)) {
            return [hoistPluralOrSelectElement(ast, el, i)];
        }
        if (isTagElement(el) && findPluralOrSelectElement([el])) {
            throw new Error('Cannot hoist plural/select within a tag element. Please put the tag element inside each plural/select option');
        }
    }
    return ast;
}
function isStructurallySamePluralOrSelect(el1, el2) {
    var options1 = el1.options;
    var options2 = el2.options;
    if (Object.keys(options1).length !== Object.keys(options2).length) {
        return false;
    }
    for (var key in options1) {
        if (!options2[key]) {
            return false;
        }
        if (!isStructurallySame(options1[key].value, options2[key].value)) {
            return false;
        }
    }
    return true;
}
export function isStructurallySame(a, b) {
    var aWithoutLiteral = a.filter(function (el) { return !isLiteralElement(el); });
    var bWithoutLiteral = b.filter(function (el) { return !isLiteralElement(el); });
    if (aWithoutLiteral.length !== bWithoutLiteral.length) {
        return false;
    }
    var elementsMapInA = aWithoutLiteral.reduce(function (all, el) {
        if (isPoundElement(el)) {
            all['#'] = el;
            return all;
        }
        all[el.value] = el;
        return all;
    }, {});
    var elementsMapInB = bWithoutLiteral.reduce(function (all, el) {
        if (isPoundElement(el)) {
            all['#'] = el;
            return all;
        }
        all[el.value] = el;
        return all;
    }, {});
    for (var _i = 0, _a = Object.keys(elementsMapInA); _i < _a.length; _i++) {
        var varName = _a[_i];
        var elA = elementsMapInA[varName];
        var elB = elementsMapInB[varName];
        if (!elB) {
            return false;
        }
        if (elA.type !== elB.type) {
            return false;
        }
        if (isLiteralElement(elA) || isLiteralElement(elB)) {
            continue;
        }
        if (isArgumentElement(elA) &&
            isArgumentElement(elB) &&
            elA.value !== elB.value) {
            return false;
        }
        if (isPoundElement(elA) || isPoundElement(elB)) {
            continue;
        }
        if (isDateElement(elA) ||
            isTimeElement(elA) ||
            isNumberElement(elA) ||
            isDateElement(elB) ||
            isTimeElement(elB) ||
            isNumberElement(elB)) {
            if (elA.value !== elB.value) {
                return false;
            }
        }
        if (isPluralElement(elA) &&
            isPluralElement(elB) &&
            !isStructurallySamePluralOrSelect(elA, elB)) {
            return false;
        }
        if (isSelectElement(elA) &&
            isSelectElement(elB) &&
            isStructurallySamePluralOrSelect(elA, elB)) {
            return false;
        }
        if (isTagElement(elA) && isTagElement(elB)) {
            if (elA.value !== elB.value) {
                return false;
            }
            if (!isStructurallySame(elA.children, elB.children)) {
                return false;
            }
        }
    }
    return true;
}
